SourceKitStressTester/Sources/StressTester/SourceKitDocument.swift

```diff
@@ -42,29 +48,52 @@ struct SourceKitDocument {
     let needsLineColumn: Bool = true
   }
 
-  init(_ file: String, args: [String], connection: SourceKitdService, containsErrors: Bool = false) {
+  init(_ file: URL, swiftc: String, args: [String],
+       tempDir: URL, connection: SourceKitdService,
+       containsErrors: Bool = false) {
     self.file = file
+    self.swiftc = swiftc
     self.args = args
-    self.containsErrors = containsErrors
+    self.tempDir = tempDir
     self.connection = connection
+    self.containsErrors = containsErrors
     self.diagEngine.addConsumer(EmptyDiagConsumer())
   }
 
-  mutating func open(state: SourceState? = nil) throws -> (SourceFileSyntax, SourceKitdResponse) {
+  mutating func open(rewriteMode: RewriteMode) throws -> (SourceFileSyntax, SourceKitdResponse) {
+    return try openOrUpdate(rewriteMode: rewriteMode)
+  }
+
+  // FIXME: Can we not have the escaping here? It clearly doesn't since the
+  //        function it's calling doesn't escape?
+  mutating func update(updateSource: @escaping (inout SourceState) -> Void) throws -> (SourceFileSyntax, SourceKitdResponse) {
+    return try openOrUpdate(updateSource: updateSource)
+  }
+
+  private mutating func openOrUpdate(rewriteMode: RewriteMode? = nil,
+                                     updateSource: ((inout SourceState) -> Void)? = nil)
+  throws -> (SourceFileSyntax, SourceKitdResponse) {
     let request = SourceKitdRequest(uid: .request_EditorOpen)
-    if let state = state {
-      sourceState = state
-      request.addParameter(.key_SourceText, value: state.source)
-    } else {
-      request.addParameter(.key_SourceFile, value: file)
+
+    if let rewriteMode = rewriteMode {
+      let source = try! String(contentsOf: file, encoding: .utf8)
+      sourceState = SourceState(rewriteMode: rewriteMode, content: source)
+
+      request.addParameter(.key_SourceFile, value: file.path)
+    } else if let updateSource = updateSource {
+      var sourceState = self.sourceState!
+      _ = try close()
```

Comment:
I'm not sure we're ever checking close's return value. Maybe we should add `@discardableResult` on it so we can drop the underscores?

Reply:
Sounds good. You wouldn't happen to know if I can remove the @escaping somehow as well would you? I'm not a huge fan of those three functions but I'm not sure how to do it another way while also avoiding duplicating most of open/update.

Reply:
I think it's because the function being called wraps it in an optional, which stores it, and the analysis gives up at that point. I think the only way to avoid it is to not have that optional. You could make it two separate overloads, rather than an optional param (which seems like it would defeat the purpose) or make it non-optional but default to closure that does nothing `{}`. Maybe it's better to pull the `sourceState` handling bits out from the `if` into their respective callers and pass in optional source text or file path instead?

Reply:
Much nicer, thanks!